flaky-test: HeatmapGrid.test.tsx — 5 tests fail intermittently with 'expected 1 to be greater than 1'

## Summary

`src/__tests__/HeatmapGrid.test.tsx` has 5 tests that fail intermittently in CI with the same assertion error:

```
AssertionError: expected 1 to be greater than 1
```

### Failing tests
- applies aria-labels with date and value to cells
- uses color intensity based on value relative to max
- formats tooltip values using formatValue prop
- shows tooltip on cell hover
- hides tooltip on cell leave

### Evidence
- Seen on PR #2839 (Dependabot lockfile bump — no code changes) in both `dashboard-test` and `test (ubuntu-latest, 20)` jobs
- Same commit passes on `develop` branch CI — confirming this is a flake, not a regression
- `Test Files: 1 failed | 85 passed (86)` — only this file fails
- Likely a DOM query/timing issue in jsdom where elements are not fully rendered before assertion

### Impact
- Blocks PRs from getting clean CI (false negatives)
- Wastes reviewer time investigating non-regressions

### Suggested fix
- Add `waitFor` or `findBy` queries instead of direct `querySelectorAll`
- Ensure the HeatmapGrid component is fully rendered before asserting DOM structure
- Consider adding `vi.useFakeTimers()` if there are animation/frame dependencies

### Priority
P2 — not production-breaking but CI noise that masks real failures

--- a/dashboard/src/__tests__/HeatmapGrid.test.tsx
+++ b/dashboard/src/__tests__/HeatmapGrid.test.tsx
@@ -1,22 +1,47 @@
 /**
  * __tests__/HeatmapGrid.test.tsx
+ *
+ * Fixed flaky assertions (issue #2848):
+ * - Replaced Math.random() with deterministic seed data
+ * - Scoped all DOM queries to rendered container (not global document)
+ * - Fixed month label test to use fake timer date consistently
  */
 
 import { afterEach, beforeEach, describe, it, expect, vi } from 'vitest';
 import { render, screen, fireEvent } from '@testing-library/react';
 import { HeatmapGrid, type HeatmapDataPoint } from '../components/analytics/HeatmapGrid';
 
-/** Generate daily data for the last N days. */
-function generateDailyData(days: number, maxVal: number = 100): HeatmapDataPoint[] {
+/** Generate deterministic daily data for the last N days (no Math.random). */
+function generateDailyData(days: number, startValue: number = 1): HeatmapDataPoint[] {
   const data: HeatmapDataPoint[] = [];
-  const now = new Date();
+  // Use the fake timer's fixed time (2026-04-28)
+  const now = new Date('2026-04-28T12:00:00Z');
   for (let i = 0; i < days; i++) {
     const d = new Date(now);
     d.setDate(d.getDate() - i);
     const dateStr = d.toISOString().split('T')[0] ?? '';
+    // Deterministic: spread values across the range
     data.push({
       date: dateStr,
-      value: Math.round(Math.random() * maxVal),
+      value: startValue + (i * 7) % 100,
+    });
+  }
+  return data;
+}
+
+/** Generate data with guaranteed non-zero values for color tests. */
+function generateColorfulData(days: number): HeatmapDataPoint[] {
+  const data: HeatmapDataPoint[] = [];
+  const now = new Date('2026-04-28T12:00:00Z');
+  for (let i = 0; i < days; i++) {
+    const d = new Date(now);
+    d.setDate(d.getDate() - i);
+    const dateStr = d.toISOString().split('T')[0] ?? '';
+    // Deterministic spread across intensity levels
+    const levelValues = [0, 10, 50, 75, 100];
+    data.push({
+      date: dateStr,
+      value: levelValues[i % levelValues.length],
     });
   }
   return data;
@@ -81,7 +106,8 @@ describe('HeatmapGrid', () => {
   });
 
   it('supports different color scales', () => {
-    const data = generateDailyData(14);
+    // Use deterministic data with guaranteed non-zero values
+    const data = generateColorfulData(14);
     const { container: cyanContainer } = render(
       <HeatmapGrid data={data} weeks={2} color="cyan" />,
     );
@@ -128,24 +154,26 @@ describe('HeatmapGrid', () => {
     expect(screen.getByText('Less')).toBeTruthy();
     expect(screen.getByText('More')).toBeTruthy();
 
-    // Legend should have 5 color swatches
+    // Legend should have 5 color swatches — scoped to container
     const swatches = container.querySelectorAll('.rounded-sm');
     expect(swatches).toHaveLength(5);
   });
 
   it('renders month labels for transitions', () => {
-    // Generate data spanning 2 months
+    // Generate data spanning 2 months using fixed fake-timer date
     const data: HeatmapDataPoint[] = [];
+    const baseDate = new Date('2026-04-28T12:00:00Z');
     for (let i = 0; i < 60; i++) {
-      const d = new Date();
+      const d = new Date(baseDate);
       d.setDate(d.getDate() - i);
       data.push({ date: d.toISOString().split('T')[0] ?? '', value: 10 });
     }
-    render(<HeatmapGrid data={data} weeks={9} />);
-    // Should have at least 1 month label
-    const textElements = document.querySelectorAll('svg text');
+    const { container } = render(<HeatmapGrid data={data} weeks={9} />);
+    // Should have at least 1 month label — scoped to container, not global document
+    const textElements = container.querySelectorAll('svg text');
+    const MONTH_NAMES = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'];
     const monthTexts = Array.from(textElements).filter(
-      (t) => t.textContent && ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'].includes(t.textContent),
+      (t) => t.textContent && MONTH_NAMES.includes(t.textContent),
     );
     expect(monthTexts.length).toBeGreaterThanOrEqual(1);
   });
@@ -155,14 +183,13 @@ describe('HeatmapGrid', () => {
       { date: '2026-04-28', value: 99 },
       { date: '2026-04-27', value: 0 },
     ];
-    render(<HeatmapGrid data={data} weeks={1} />);
+    const { container } = render(<HeatmapGrid data={data} weeks={1} />);
 
     const cell = screen.getByRole('gridcell', { name: /2026-04-28/ });
     fireEvent.mouseEnter(cell);
 
-    // Tooltip should appear as SVG text
-    const svg = screen.getByRole('img');
-    const tooltipGroup = svg.querySelector('g[pointer-events="none"]');
+    // Tooltip should appear as SVG text — scoped to container
+    const tooltipGroup = container.querySelector('g[pointer-events="none"]');
     expect(tooltipGroup).toBeTruthy();
   });
 
@@ -170,15 +197,14 @@ describe('HeatmapGrid', () => {
     const data: HeatmapDataPoint[] = [
       { date: '2026-04-28', value: 50 },
     ];
-    render(<HeatmapGrid data={data} weeks={1} />);
+    const { container } = render(<HeatmapGrid data={data} weeks={1} />);
 
     const cell = screen.getByRole('gridcell', { name: /2026-04-28/ });
     fireEvent.mouseEnter(cell);
     fireEvent.mouseLeave(cell);
 
-    // Tooltip group should be gone
-    const svg = screen.getByRole('img');
-    const tooltipGroup = svg.querySelector('g[pointer-events="none"]');
+    // Tooltip group should be gone — scoped to container
+    const tooltipGroup = container.querySelector('g[pointer-events="none"]');
     expect(tooltipGroup).toBeNull();
   });
 

